### ISSUE TITLE
[X86] `sub`+`cmp`+`jmp` could be combined into `sub`+`jmp`

### ISSUE THREAD

**Issue by chfast** [llvm/llvm-project#58636]
```c++
void work(int);

void test(int g)
{
    g -= 3;
    if (g < 0) [[unlikely]]
        return;
    work(g);
}
```
produces
```asm
test(int):
        cmp     edi, 3
        jl      .LBB0_1
        add     edi, -3
        jmp     work(int)
.LBB0_1:
        ret
```
but the better would be (as done by GCC)
```asm
test(int):
        sub     edi, 3
        js      .L1
        jmp     work(int)
.L1:
        ret
```

https://godbolt.org/z/Knhn3jP8n

**Comment by davidbolvansky** [comment]
Regressed since llvm 10.

**Comment by chfast** [comment]
Workaround:
```cpp
void workaround(int g)
{
    unsigned h;
    bool o = __builtin_usub_overflow(g, 3, &h);
    if (o) [[unlikely]]
        return;
    work(h);
}
```
https://godbolt.org/z/MjWe91cxc

**Comment by e-kud** [comment]
As @davidbolvansky noticed it is a regression since llvm10. I bisected it to commit https://github.com/llvm/llvm-project/commit/0f22e783a038b6983f0fe161eef6cf2add3a4156

Apparently this change has some mystery. 

> As i have briefly discusses in IRC with Tim, the original motivation
can no longer be recovered, too much time has passed.
>
> However i believe that the original fold was doing the right thing,
we should be performing such a transformation even if the inner `add`
will not go away - that will still unchain the comparison from `add`,
it will no longer need to wait for `add` to compute.

This change combines `cmp` + `add` even if `add` has multiple users.
```llvm
; Function Attrs: mustprogress uwtable
define dso_local void @_Z4testi(i32 noundef %0) local_unnamed_addr #0 {
  %2 = add nsw i32 %0, 3
  %3 = icmp slt i32 %2, 0
  %4 = zext i1 %3 to i64
  br i1 %3, label %6, label %5, !prof !6

5:                                                ; preds = %1
  call void @_Z4worki(i32 noundef %2)
  br label %6

6:                                                ; preds = %1, %5
  ret void
}
*** IR Dump After InstCombinePass on _Z4testi ***
; Function Attrs: mustprogress uwtable
define dso_local void @_Z4testi(i32 noundef %0) local_unnamed_addr #0 {
  %2 = icmp slt i32 %0, -3
  br i1 %2, label %5, label %3, !prof !6

3:                                                ; preds = %1
  %4 = add nsw i32 %0, 3
  call void @_Z4worki(i32 noundef %4)
  br label %5

5:                                                ; preds = %1, %3
  ret void
}
```
Then `icmp` and `add` are lowered into separate `cmp     edi, 3` and `add     edi, -3` as it is shown in the subject. It seems to me that codegen is too late to fix it – the original `add` has already been sunk into `if` block and we need to do some hoisting. That's why I decided to address to the original patch.

I think it works good for the cases as shown below. We have a comparison with a value and there is no much difference with which value we compare `100` or `92`.
```llvm
%tmp4.7 = add nuw nsw i64 %tmp3, 8
%tmp6.7 = icmp ult i64 %tmp4.7, 100
```
In case of comparison with non-zero, we impose extra arithmetic. However, many targets have an ability to compare effectively with zero after math operations. I propose to limit the original patch to comparisons with non-zero values. [D144764](https://reviews.llvm.org/D144764).

It changes `X86` as follows (even better than GCC):
```asm
# Before D144764
test(int):                               # @test(int)
        cmp     edi, -3
        jl      .LBB0_1
        add     edi, 3
        jmp     work(int)@PLT                    # TAILCALL
.LBB0_1:
        ret

# After D144764
test(int):                               # @test(int)
        addl    $-3, %edi                                  
        jns     work(int)@PLT                     # TAILCALL
        retq                           
```
For `AArch64` we need to change `g -= 3` -> `g += 3` to see the effect.
```asm
// Before D144764
test(int):                               // @test(int)
        cmn     w0, #3
        b.lt    .LBB0_2
        add     w0, w0, #3
        b       work(int)
.LBB0_2:
        ret

// After D144764
test(int):                               // @test(int)
        adds    w0, w0, #3                         
        b.mi    .LBB0_2                            
        b       _Z4worki                           
.LBB0_2:
        ret                                        
```
@LebedevRI what do you think?

**Comment by nikic** [comment]
There already is a backend undo fold for this at https://github.com/llvm/llvm-project/blob/7910ed1d56c349b76c82d5ebe2f2590770955ff5/llvm/lib/CodeGen/CodeGenPrepare.cpp#L7956

For `add`, it's currently limited to equality comparisons though, so it doesn't handle this precise pattern yet.

**Comment by chfast** [comment]
@LebedevRI any update here? I can try to address in CodeGenPrepare as pointed by @nikic but there is also a fix draft by @e-kud in [D144764](https://reviews.llvm.org/D144764).

**Comment by nikic** [comment]
If the backend undo fold can recover this, we'll want to do that. If not, we can start considering InstCombine limitations.

**Comment by e-kud** [comment]
Sorry, I haven't thought that there is still the interest to it.

I've made the change suggested by @niktesic. It is here https://reviews.llvm.org/D145454. It works fine but it revealed a problem (my previous approach has the same problem, but it is hidden from tests).

Consider the following code.
```
void decref(long* refcount) {
   long count = *refcount;
   if (count == 1) { free_object() }
   else if (count > 1) { *refcount = count - 1; }
   else { /* immortal */ }
}
```
Here we can make only one comparison of `count` against `1` and use it for both branches. This optimization is implemented.

When we move to usage of flags from `add` or `sub` we have the following situation:
```
    decl %eax
    je .LBB0_3
# %bb.1: # %bb2
    testl %eax, %eax
    jle .LBB0_4
```
And generally we can't eliminate `test` because of `decl` because `test` resets `CF` flag in contrast to `add`. The bummer here is that `test` appeared from `CMP, 0`, but what if it isn't?

I tried to use `hasNoSignedWrap` predicate but it is not propagated to `CodeGen`.

This is where I stuck, any ideas?

**Comment by zmodem** [comment]
I was looking at this the other day and made some notes:

> And generally we can't eliminate `test` because of `decl` because `test` resets `CF` flag in contrast to `add`.

Do you mean we can't eliminate `test` because `decl` *doesn't set* the `CF` flag?

I think we don't need `CF` in this case. `JLE` checks `ZF`, `SF`, and `OF` -- which should be set by `decl`. So I think this should work:

```
    decl %eax
    je .LBB0_3
# %bb.1: # %bb2
    jle .LBB0_4
```

We'd need to teach `X86InstrInfo::optimizeCompareInstr` to reuse those flags though. Maybe that would be a worthwhile improvement in itself?

---

It seems like an oversight that `preferZeroCompareBranch()` isn't already true for X86. (It was initially added for ARM in dd5c52029d27125bffaca1d1d442eedeae6481b9.) We should probably try enabling that regardless of this issue.

---

The "preferAnyZeroCompareBranch()" approach in https://reviews.llvm.org/D145454 only works for *signed* comparisons. I.e.

```
%cmp = icmp ult i32 %i, 42
br i1 %cmp, ...
%sub = add i32 %i, -42
```

to

```
%sub = add i32 %i, -42
%0 = icmp ult i32 %sub, 0
br i1 %0, ...
```

is not valid, because unlike the signed case we don't want to check whether `%i - 42 < 0` (always true for unsigned) -- we'd want to check whether `%i - 42` overflows.

We'd need to find a different approach for handling the unsigned case.

---

fa2a6d68c6f35e478af3425760ae39e866049eee added support for `XOR` in addition to `ADD` and `SUB` in this `.optimizeBranch()` code. That only works for equality comparisons.

**Comment by e-kud** [comment]
> I think we don't need CF in this case. JLE checks ZF, SF, and OF -- which should be set by decl. So I think this should work:

Good point. I think at that moment, I was thinking that `test` and `dec` set different things, and could be removed, but we also need to watch on all the eflags users of `test` in this case. Because it is also possible that after `jle` there is another user of EFLAGS and it may use `CF`. Probably we can just check if EFLAGS dead on the jump instructions.

> is not valid, because unlike the signed case we don't want to check whether %i - 42 < 0 (always true for unsigned) -- we'd want to check whether %i - 42 overflows.

Agree. But I'm not sure if this condition survives IR optimizations, does it?
